### PR TITLE
Epoch.Chart.Canvas#redraw() method

### LINKED ISSUE #163: API to Redraw Canvas Based Charts
It would be great to have an API to re-render Canvas based charts to handle on demand changes such as color scheme changes and window resizing.

Think of a use case where users are allow to switch "themes" (change of CSS rules that are initially applied to the charts) of a page that contains some basic charts, and some real time ones. By switching "themes" users should be able to see the changes happen to both types of charts. Currently only the pure SVG (basic) ones reflect the immediate CSS changes (I guess due to SVG's nature).

In such use cases, an API to allow on-demand redraw of the Canvas based charts can be very useful.

### PR DESCRIPTION
**Pull request by woozyking**
Fixes #163

### CODE REVIEW

**Comment by woozyking** [comment]
I'm not entirely sure how to write the test for this yet, consider this WIP, need a bit of mentoring from @rsandor

**Comment by rsandor** [comment]
Okay so if you want to tackle this you'll need to use a couple of spies! We should really be doing this in more places but we are not (since I was not the unit testing guru I am now back when I wrote this, haha). Here's roughly the idea:

**First:** From the command-line in the project run the following command: `npm install --save-dev sinon` - This will install a great JS stubbing and spying library called [sinon](http://sinonjs.org/).

**Next:** In the tests for `Epoch.Chart.Canvas` you'll need to setup [sinon](http://sinonjs.org/) spies for both of the methods you are calling from `redraw`. Here's an example of how it might look:

Around line 1 in `tests/unit/core/charts.coffee` you'll need to require the sinon library in the test:
```coffeescript
sinon = require 'sinon'
```

**Then:** you'll need to create a new "describe" block for the redraw method. The tests for `Epoch.Chart.Canvas` start around line 400. You should see multiple "describe" blocks: first is for the constructor, then another one for `getWidth`, etc. Find a good place somewhere in there and add your own new describe block, something like this:

```coffeescript
describe 'redraw', ->
  chart = null
  # This means: before you run this entire suite of tests, create a new canvas chart
  before ->
    chart = new Epoch.Chart.Canvas()
  # Here it means: before each individual test, set a spy on the chart's draw method
  # and also set a spy on the Epoch.QueryCSS.purge method
  beforeEach ->
    sinon.spy chart, 'draw'
    sinon.spy Epoch.QueryCSS, 'purge'
  # This is saying: remove the spies after each test has been performed
  afterEach ->
    chart.draw.restore()
    Epoch.QueryCSS.purge.restore()
```

**Finally:** You'll need to write the individual tests. I leave this as an exercise for you to figure out how to do (since it's relatively easy). A minor hint though: `assert.ok chart.draw.calledOnce`.

Happy testing :dancers:

File: src/core/chart.coffee [hunk 1]
```diff
@@ -327,3 +327,8 @@ class Epoch.Chart.Canvas extends Epoch.Chart.Base
     super()
     @canvas.style {'width': "#{@width}px", 'height': "#{@height}px"}
     @canvas.attr { width: @getWidth(), height: @getHeight() }
+
+  # Purges QueryCSS cache and redraws the Canvas based chart.
+  redraw: ->
```
Comment (rsandor) [inline src/core/chart.coffee]: This is a pretty good name for this, kudos on the choice

**Comment by woozyking** [comment]
Thanks @rsandor , I'll try to come up with some valid tests for the method.

**Comment by rsandor** [comment]
@woozyking - How is this coming along?

**Comment by woozyking** [comment]
Sorry I've been distracted a bit. Will get to this tomorrow, hopefully.



—
Runzhou Li (Leo)

**Comment by rsandor** [comment]
Nice let me know if you need any help with it :)

**Comment by woozyking** [comment]
Sorry for the delay @rsandor , I'm on vacation and my access to internet/computer is irregular :smiling_imp: 

I'm not certain what more tests should be done on `redraw` method since it basically invokes a CSS cache purge and then a `draw`.

If I understand correctly, `sinon` here is used to see whether those spied methods are called properly, hence reusing (?please confirm) the test cases written for those spied methods. Therefore I only added one test case for it asserting on the fact that those two methods are each called exactly once. On a side note, I really like `sinon`'s concept, wish I'd known it earlier.

Let me know what other tests can be done on `redraw`.

**Comment by rsandor** [comment]
Yep that's it, merging. :100:

**Comment by woozyking** [comment]
:laughing: nice, thanks.

**Pull request merged**